### PR TITLE
New tooling for comparing the keyword dictionary and data model schemas

### PR DESCRIPTION
**Pull request by braingram**
https://jira.stsci.edu/browse/JP-3783

As it currently stands this PR:
- adds a new (private) submodule `stdatamodels.jwst._kwtool`
- with a rudimentary CLI `python -m stdatamodels.jwst._kwtool.cli <keyword dictionary path>`
- that generates a report html file (example attached to https://jira.stsci.edu/browse/JP-3783)

Link to added docs: https://stdatamodels--337.org.readthedocs.build/en/337/jwst/kwtool/index.html

It would be helpful to have feedback on:
- What comparisons are useful?
- What format(s) are most helpful?

Some known issues are:
- "enum" comparison will need to account for the datamodel schemas allowing now-unused values. These are currently reported as differences but being able to specify an allowed difference (which still checking the other values in the enum) would be useful to de-clutter the report.
- "path" differences are reported for keywords that are re-used in multi-model models. For example EXP_TYPE currently reports a "path" difference with "meta.exposure.type" in "kwd" and multiple paths in "dwd" including the "meta.exposure.type" but also "trace.items.meta.exposure.type" (as one example). This is coming from https://github.com/spacetelescope/stdatamodels/blob/503320f6325f7dd454dc2d7c177043ddf069c6f6/src/stdatamodels/jwst/datamodels/schemas/spectracesingle.schema.yaml#L8 referenced here https://github.com/spacetelescope/stdatamodels/blob/503320f6325f7dd454dc2d7c177043ddf069c6f6/src/stdatamodels/jwst/datamodels/schemas/spectrace.schema.yaml#L16 (nested in `meta.items`). At first glance I would say this is a bug as I don't see how having every item in the list overwrite the PRIMARY EXP_TYPE is useful. However this might be a "bug" in the comparison where we don't care about this difference and we'll want some way to ignore it or not flag it as an issue in the first place. @tapastro does this usage look like a schema bug? If so I can open a separate issue. 

## Tasks
- [ ] update or add relevant tests
- [ ] update relevant docstrings and / or `docs/` page
- [ ] Does this PR change any API used downstream? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types)
  - [ ] [run `jwst` regression tests](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) with this branch installed (`"git+https://github.com/<fork>/stdatamodels@<branch>"`)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.feature.rst``: new feature
- ``changes/<PR#>.bugfix.rst``: fixes an issue
- ``changes/<PR#>.doc.rst``: documentation change
- ``changes/<PR#>.removal.rst``: deprecation or removal of public API
- ``changes/<PR#>.misc.rst``: infrastructure or miscellaneous change
</details>

### CODE REVIEW

File: docs/source/jwst/kwtool/keyword_dictionary.rst [hunk 1]
```diff
@@ -0,0 +1,166 @@
+.. _keyword_dictionary:
+
+Keyword Dictionary
+==================
+
+The "keyword dictionary" is a collection
+of json files that describe FITS keywords used in JWST files.
+Although the format is similar to jsonschema is is not compatible
```
Comment (chamblee-st) [inline docs/source/jwst/kwtool/keyword_dictionary.rst:8]: "is is" should be "it is"

Comment (braingram) [reply docs/source/jwst/kwtool/keyword_dictionary.rst]: Thanks! I applied that fix in b94d57d

**Comment by braingram** [comment]
Based on discussion with @tapastro there are a few remaining issues to sort out with this PR prior to approval (based on discussion of the currently generated report).

- [x] ignore "path" differences for datamodel schema keyword definitions nested under "items" arrays (like in "HDU: EXTRACT1D KEYWORD: EXTR_X"). These don't provide a mapping that would be useful for triggering reprocessing so even if they are listed in the archive they don't allow a one-to-one mapping of value-to-file. The tool will now ignore path differences if "items" is in the path. Fixed in 66092eb
- [ ] allow some enum differences (so datamodels can read old files that use outdated values and schemas don't need to redefine things like enum = {T, F} for boolean keywords)

**Comment by braingram** [comment]
@tapastro I think this is ready for final review. See the comment above https://github.com/spacetelescope/stdatamodels/pull/337#issuecomment-2445312775 for a checklist and linked commits for the items we discussed.

I also added a test that fetches the latest "published" keyword dictionary from MAST (using an "unofficial" service since there is no official one) and parses it generating a report as a test of the new tool. I think this is worth expanding by adding a CI job that:
- checks out stdatamodels main, generates a report
- checks out the PR branch, generates a report
- runs diff on the report (and updates both as artifacts)
This way we will know when PRs impact keyword dictionary differences (relative the published version, not the latest dev version which is not public). However it makes sense to me to split this work into a separate PR as I'd like to get this one in to also start working on PRs to address the many differences.

**Review by tapastro — APPROVED** [review]
🎉

File: docs/source/jwst/kwtool/keyword_dictionary.rst [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+undocumented key/value pairs. This documentation will not attempt
+to fill in these gaps but instead will focus only on the entries
+relevant to the keyword tool (for comparing keywords between this
+resource and the datamodel schemas).
+
+
+Data model paths
+----------------
+
+The heirarchical structure in which the of the keyword dictionary definitions
```
[the remark this answers is not in the document]
Comment (braingram) [reply docs/source/jwst/kwtool/keyword_dictionary.rst]: Must remind self of "i before e"... Thanks! Fixed in [616a09b](https://github.com/spacetelescope/stdatamodels/pull/337/commits/616a09b33d1af55c2c92d3ad0a6d901c60d2bc1f)

File: docs/source/jwst/kwtool/keyword_dictionary.rst [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+   :language: json
+
+Which contains a keyword definition "reference_frame" for FITS keyword
+"RADESYS". The resulting "data model path" (or "data model name") for this
+keyword definition includes the final key and is then stripped of all
+"properties" (and "allOf") keys resulting in `meta.coordinates.reference_frame`
+for the final "data model path".
+
+It is also possible for referenced files to contain additional
+heirarchical structure (objects with "properties", "allOf", etc).
```
[the remark this answers is not in the document]
Comment (braingram) [reply docs/source/jwst/kwtool/keyword_dictionary.rst]: Fixed in [616a09b](https://github.com/spacetelescope/stdatamodels/pull/337/commits/616a09b33d1af55c2c92d3ad0a6d901c60d2bc1f)

**Pull request merged**